File: web-common/src/lib/number-formatting/strategies/per-range.ts
import {
  type Formatter,
  type FormatterOptionsCommon,
  type FormatterRangeSpecsStrategy,
  type LocaleConfig,
  NumberKind,
  type NumberParts,
  type RangeFormatSpec,
} from "../humanizer-types";
import { countDigits, countNonZeroDigits } from "../utils/count-digits";
import {
  formatNumWithOrderOfMag,
  orderOfMagnitude,
  orderOfMagnitudeEng,
} from "../utils/format-with-order-of-magnitude";
import { numberPartsToString } from "../utils/number-parts-utils";
import { shortScaleSuffixIfAvailableForStr } from "../utils/short-scale-suffixes";

const formatWithRangeSpec = (
  x: number,
  spec: RangeFormatSpec,
  locale?: LocaleConfig,
): NumberParts => {
  if (spec.overrideValue !== undefined) {
    return { ...spec.overrideValue, locale };
  }
  const baseMag = spec.baseMagnitude ?? orderOfMagnitudeEng(spec.minMag);
  const padWithInsignificantZeros =
    spec.padWithInsignificantZeros === undefined
      ? true
      : spec.padWithInsignificantZeros;
  const useTrailingDot =
    spec.useTrailingDot === undefined ? true : spec.useTrailingDot;
  return formatNumWithOrderOfMag(
    x,
    baseMag,
    spec.maxDigitsRight,
    padWithInsignificantZeros,
    useTrailingDot,
    false,
    locale,
  );
};

const numberPartsValidForRangeSpec = (
  parts: NumberParts,
  spec: RangeFormatSpec,
): boolean => {
  const maxDigitsLeft = spec.maxDigitsLeft ?? 3;
  return (
    countDigits(parts.int) <= maxDigitsLeft &&
    countDigits(parts.frac) <= spec.maxDigitsRight
  );
};

const numPartsNotZero = (parts: NumberParts): boolean => {
  return (
    countNonZeroDigits(parts.int) > 0 || countNonZeroDigits(parts.frac) > 0
  );
};

export class PerRangeFormatter implements Formatter {
  options: FormatterOptionsCommon & FormatterRangeSpecsStrategy;
  initialSample: number[];
  locale?: LocaleConfig;

  constructor(
    options: FormatterRangeSpecsStrategy & FormatterOptionsCommon,
    locale?: LocaleConfig,
  ) {
    this.options = options;
    this.locale = locale;

    // sort ranges from small to large by lower bound
    this.options.rangeSpecs = this.options.rangeSpecs.sort(
      (a, b) => a.minMag - b.minMag,
    );

    // Throw an error if any of the ranges do not have min<sup
    this.options.rangeSpecs.forEach((r) => {
      if (r.minMag >= r.supMag) {
        throw new Error(
          `invalid range: min ${r.minMag} is not strictly less than sup ${r.supMag}`,
        );
      }
    });

    // Throw an error if the ranges overlap
    for (let i = 0; i < this.options.rangeSpecs.length - 1; i++) {
      const range1 = this.options.rangeSpecs[i];
      const range2 = this.options.rangeSpecs[i + 1];
      if (range1.supMag > range2.minMag) {
        throw new Error(
          `Ranges must not overlap. range 1 sup = ${range1.supMag} is greater than range 2 min = ${range2.minMag}`,
        );
      }
    }

    // Throw an error if there is a gap in coverage overlap
    for (let i = 0; i < this.options.rangeSpecs.length - 1; i++) {
      const range1 = this.options.rangeSpecs[i];
      const range2 = this.options.rangeSpecs[i + 1];
      if (range1.supMag !== range2.minMag) {
        throw new Error(
          `Gaps are not allowed between formatter ranges: range 1 sup = ${range1.supMag} is not equal to range 2 min = ${range2.minMag}`,
        );
      }
    }
  }

  stringFormat(x: number): string {
    return numberPartsToString(this.partsFormat(x));
  }

  partsFormat(x: number): NumberParts {
    const { rangeSpecs, defaultMaxDigitsRight } = this.options;

    const isPercent = this.options.numberKind === NumberKind.PERCENT;

    if (isPercent) x = 100 * x;

    let numParts: NumberParts | undefined = undefined;

    if (x === 0) {
      numParts = {
        int: "0",
        dot: "",
        frac: "",
        suffix: "",
        locale: this.locale,
      };
    }

    if (numParts === undefined) {
      // Strategy: try to format the number with each spec
      // from smallest OoM to largest, and see whether that
      // range's rules are satisfied and result in a non-zero
      // formatted number.
      for (let i = 0; i < rangeSpecs.length; i++) {
        const spec = rangeSpecs[i];
        if (spec.overrideValue !== undefined) {
          const OoM = orderOfMagnitude(x);
          if (OoM >= spec.minMag && OoM < spec.supMag) {
            numParts = { ...spec.overrideValue, locale: this.locale };
            break;
          }
        }
        numParts = formatWithRangeSpec(x, spec, this.locale);
        if (
          numberPartsValidForRangeSpec(numParts, spec) &&
          numPartsNotZero(numParts)
        ) {
          // if we have successfully formatted with this spec, we're done...
          break;
        } else {
          // Set numparts back to undefined so and continue
          // through the loop to the next step. If this was the
          // final spec, we'll fall back to the default after
          //  exiting the loop
          numParts = undefined;
        }
      }
    }

    // if numParts is still undefined after trying all specs,
    // use defaults
    if (numParts === undefined) {
      const magE = orderOfMagnitudeEng(x);
      const hasSmallFraction = magE <= -3;

      const maxDigitsLeft = hasSmallFraction ? 1 : 3;
      numParts = formatNumWithOrderOfMag(
        x,
        magE,
        defaultMaxDigitsRight,
        true,
        false,
        false,
        this.locale,
      );
      // Note that if this attempt at formatting results in more
      // digits left of the decimal point than maxDigitsLeft, then we must format this
      // number according to the next magnitude up. We'll attempt this up to 3 times
      // to find a suitable formatting, incrementing the magnitude each time.
      let attempts = 0;
      while (countDigits(numParts.int) > maxDigitsLeft && attempts < 3) {
        numParts = formatNumWithOrderOfMag(
          x,
          magE + maxDigitsLeft + attempts,
          defaultMaxDigitsRight,
          true,
          false,
          false,
          this.locale,
        );
        attempts++;
      }
    }

    numParts.suffix = shortScaleSuffixIfAvailableForStr(numParts.suffix);

    if (this.options.upperCaseEForExponent !== true) {
      numParts.suffix = numParts.suffix.replace("E", "e");
    }
    if (this.options.numberKind === NumberKind.DOLLAR) {
      numParts.currencySymbol = "$";
    } else if (this.options.numberKind === NumberKind.EURO) {
      numParts.currencySymbol = "€";
    }
    if (this.options.numberKind === NumberKind.PERCENT) {
      numParts.percent = "%";
    }

    return numParts;
  }
}

File: runtime/pkg/formatter/per_range.go
package formatter

import (
	"errors"
	"fmt"
	"sort"
)

type formatterRangeSpecsStrategy struct {
	numKind               numberKind
	upperCaseEForExponent bool
	defaultMaxDigitsRight int
	rangeSpecs            []rangeFormatSpec
}

type perRangeFormatter struct {
	options formatterRangeSpecsStrategy
}

func newPerRangeFormatter(options formatterRangeSpecsStrategy) (*perRangeFormatter, error) {
	// Sort range specs from small to large by lower bound
	sort.Slice(options.rangeSpecs, func(i, j int) bool {
		return options.rangeSpecs[i].minMag < options.rangeSpecs[j].minMag
	})

	// Validate range specs
	for i, r := range options.rangeSpecs {
		if r.minMag >= r.supMag {
			return nil, fmt.Errorf("invalid range: min %d is not strictly less than sup %d", r.minMag, r.supMag)
		}
		if i > 0 && options.rangeSpecs[i-1].supMag > r.minMag {
			return nil, errors.New("ranges must not overlap")
		}
		if i > 0 && options.rangeSpecs[i-1].supMag != r.minMag {
			return nil, errors.New("gaps are not allowed between formatter ranges")
		}
	}

	return &perRangeFormatter{options: options}, nil
}

func (f *perRangeFormatter) StringFormat(x any) (string, error) {
	var numberParts *numberParts
	if v, ok := asUnsigned(x); ok {
		numberParts = partsFormat(v, f.options)
		return numberParts.string(), nil
	}

	if v, ok := asInteger(x); ok {
		numberParts = partsFormat(v, f.options)
		return numberParts.string(), nil
	}

	if v, ok := asFloat(x); ok {
		numberParts = partsFormat(v, f.options)
		return numberParts.string(), nil
	}

	return "", fmt.Errorf("not a number: %v", x)
}

func partsFormat[T number](x T, ops formatterRangeSpecsStrategy) *numberParts {
	rangeSpecs := ops.rangeSpecs
	defaultMaxDigitsRight := ops.defaultMaxDigitsRight
	// Scale value for percentage if applicable
	if ops.numKind == numPercent {
		x *= 100
	}

	var numParts *numberParts
	if x == 0 {
		numParts = &numberParts{integer: "0"}
	}

	if numParts == nil {
		for _, spec := range rangeSpecs {
			np := formatWithRangeSpec(x, spec)
			if numberPartsValidForRangeSpec(np, spec) && numPartsNotZero(np) {
				numParts = &np
				break
			}
		}
	}

	// If no valid format was found, apply default formatting
	if numParts == nil {
		magE := orderOfMagnitudeEng(x)
		np := formatNumWithOrderOfMag(x, formatNumWithOrderOfMagOps{magE, defaultMaxDigitsRight, true, false, false})
		if countDigits(np.integer) > 3 {
			np = formatNumWithOrderOfMag(x, formatNumWithOrderOfMagOps{magE + 3, defaultMaxDigitsRight, true, false, false})
		}
		numParts = &np
	}

	numParts.adjustSuffix(ops.upperCaseEForExponent)

	// Adjust currency symbols and percent signs based on number kind
	switch ops.numKind {
	case numDollar:
		numParts.currencySymbol = "$"
	case numEuro:
		numParts.currencySymbol = "€"
	case numPercent:
		numParts.percent = "%"
	}

	return numParts
}

func formatWithRangeSpec[T number](x T, spec rangeFormatSpec) numberParts {
	padWithInsignificantZeros := spec.padWithInsignificantZeros
	useTrailingDot := spec.useTrailingDot

	ops := formatNumWithOrderOfMagOps{spec.baseMagnitude, spec.maxDigitsRight, padWithInsignificantZeros, useTrailingDot, false}
	return formatNumWithOrderOfMag(x, ops)
}

// numberPartsValidForRangeSpec checks if the given number parts are valid according to the specified range spec.
func numberPartsValidForRangeSpec(parts numberParts, spec rangeFormatSpec) bool {
	return countDigits(parts.integer) <= spec.maxDigitsLeft &&
		countDigits(parts.frac) <= spec.maxDigitsRight
}

// numPartsNotZero checks if the number parts represent a non-zero number.
func numPartsNotZero(parts numberParts) bool {
	return countNonZeroDigits(parts.integer) > 0 || countNonZeroDigits(parts.frac) > 0
}
